### PR TITLE
[QUIC] API changes follow up

### LINKED ISSUE #72027: [QUIC] QuicConnection API follow up
- [ ] it should AFAIK. It would be nice to have some assert about it. It differs among platforms but MsQUIC now uses native layout e.g. same as SocketAddress.

_Originally posted by @wfurt in https://github.com/dotnet/runtime/pull/71783#discussion_r919126830_

- [ ] It would be nice to throw or at least log. I know this is what we we in SslStream for historical reasons but the failures when we silently ignore given certificate are annoying IMHO. 
(this could be separate work)

_Originally posted by @wfurt in https://github.com/dotnet/runtime/pull/71783#discussion_r919132734_

### PR DESCRIPTION
**Pull request by ManickaP**
- `ResettableValueTaskSource` now propagates exceptions immediately from `GetResult`. I.e.: if it gets set successfully without `final` (via e.g. RECEIVE) and then with exception with `final` (via `AbortRead`) it'll report the exception from the abort. cc @CarnaViire 
- Added comments for `QuicStream` and both task sources
- Added some more tests (for `RemoteCertificate`, for `Task` properties of `QuicStream`)

Fixes #72027

### CODE REVIEW

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ResettableValueTaskSource.cs [hunk 1]
```diff
@@ -207,75 +244,31 @@ void IValueTaskSource.GetResult(short token)
...
+                        if (_finalTaskSource.Task.IsCompletedSuccessfully)
                         {
-                            _valueTaskSource.SetException(exception);
+                            _valueTaskSource.SetResult(true);
                         }
                         else
                         {
-                            _valueTaskSource.SetResult(true);
+                            // We know it's always going to be a single exception since we're the ones setting it.
+                            _valueTaskSource.SetException(_finalTaskSource.Task.Exception?.InnerException!);
```
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ResettableValueTaskSource.cs:260]: Why inner exception? 🤔 Above we were just setting  `_finalTaskSource.TrySetException(exception)` to whatever exception was passed to `ResettableValueTaskSource.TrySetException(exception, final: true)`

Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ResettableValueTaskSource.cs:260]: `_finalTaskSource.Task.Exception` is `AggregateException`, so I'm pulling the original exception from it via `InnerException`.

**Review by CarnaViire — APPROVED** [review]
LGTM, thanks!

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ResettableValueTaskSource.cs [hunk 2]
```diff
@@ -15,7 +15,8 @@ internal sealed class ResettableValueTaskSource : IValueTaskSource
 {
     // None -> [TryGetValueTask] -> Awaiting -> [TrySetResult|TrySetException(final: false)] -> Ready -> [GetResult] -> None
     // None -> [TrySetResult|TrySetException(final: false)] -> Ready -> [TryGetValueTask] -> [GetResult] -> None
-    // None|Awaiting -> [TrySetResult|TrySetException(final: true)] -> Final(never leaves this state)
+    // None|Awaiting -> [TrySetResult|TrySetException(final: true)] -> Completed(never leaves this state)
+    // Ready -> [GetResult: TrySet*(final: true) was called] -> Completed(never leaves this state)
```
[the remark this answers is not in the document]
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ResettableValueTaskSource.cs:19]: | task | | | | |
| - | - | - | - | - |
| _valueTaskSource | success | fail E1 | success | fail E1 |
| _finalTaskSource | success | success | fail E2 | fail E2 |
| GetResult  | success | fail E1 | fail E2 | fail E1 |
| GetResult subsequent  | success | success | fail E2 | fail E2 |

**Pull request merged**